onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,420 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "trt_execution_provider.h"
+#include "trt_allocator.h"
+#include "core/framework/execution_provider.h"
+#include "core/framework/op_kernel.h"
+#include "core/framework/kernel_registry.h"
+#include "core/framework/compute_capability.h"
+#include "core/providers/cpu/cpu_execution_provider.h"
+#include "core/platform/env.h"
+#include "onnx/shape_inference/implementation.h"
+#include "cuda_runtime_api.h"
+#include "gsl/pointers"
+#include "core/graph/model.h"
+#include "cuda_runtime_api.h"
+
+using namespace std;
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::logging;
+
+namespace onnxruntime {
+
+TRTExecutionProvider::TRTExecutionProvider()
+    : IExecutionProvider{onnxruntime::kTRTExecutionProvider} {
+  DeviceAllocatorRegistrationInfo trt_device_info({OrtMemTypeCPU, [](int) {
+                                                     return std::make_unique<TRTPinnedAllocator>();
+                                                   },
+                                                   std::numeric_limits<size_t>::max()});
+  InsertAllocator(CreateAllocator(trt_device_info));
+  DeviceAllocatorRegistrationInfo default_device_info({OrtMemTypeDefault, [](int) {
+                                                         return std::make_unique<TRTAllocator>();
+                                                       },
+                                                       std::numeric_limits<size_t>::max()});
+  InsertAllocator(CreateAllocator(default_device_info));
+}
+
+TRTExecutionProvider::~TRTExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+TRTExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
+                                    const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  // Construct modelproto from graph
+  onnxruntime::Model model(graph.Name(), true, ModelMetaData(), IOnnxRuntimeOpSchemaRegistryList(), graph.DomainToVersionMap());
+  onnxruntime::Graph& graph_build = model.MainGraph();
```

Comment:
Would 'graph' have come from a ModelProto originally and we're doing work to re-create something that already exists? If so, is there a better way to do this?

Reply:
In current GetCapability API, graph is GraphViewer object, which is read-only and doens't have the ability to convert to ModelProto. That's why here we reconstruct graph. Ideal way is to just send GraphViewer object to TensorRT and avoid converting ModelProto to Graph then back to ModelProto. But after a couple of round of discussions, TensorRT insists to use ModelProto as its input.